### PR TITLE
[#157768307] Check if we can remove Autorest

### PR DESCRIPTION
**Pull request by lussoluca**
A lot of tests will fail!

For PagoPA proxy we still use Autorest at the moment.

The node-fetch package uses types not compatibile with the ones in italia-ts-commons and I don't know how to fix them (@see `src/clients/api.ts::220`).

```
src/clients/api.ts:220:7 - error TS2345: Argument of type '{ baseUrl: string; fetchApi: (url: string | Request, init?: RequestInit | undefined) => Promise<R...' is not assignable to parameter of type '{ readonly baseUrl?: string | undefined; readonly fetchApi?: ((input: string | Request, init?: Re...'.
  Type '{ baseUrl: string; fetchApi: (url: string | Request, init?: RequestInit | undefined) => Promise<R...' is not assignable to type '{ readonly baseUrl?: string | undefined; readonly fetchApi?: ((input: string | Request, init?: Re...'.
    Types of property 'fetchApi' are incompatible.
      Type '(url: string | Request, init?: RequestInit | undefined) => Promise<Response>' is not assignable to type '((input: string | Request, init?: RequestInit | undefined) => Promise<Response>) | undefined'.
        Type '(url: string | Request, init?: RequestInit | undefined) => Promise<Response>' is not assignable to type '(input: string | Request, init?: RequestInit | undefined) => Promise<Response>'.
          Types of parameters 'url' and 'input' are incompatible.
            Type 'string | Request' is not assignable to type 'string | import("/usr/src/app/node_modules/@types/node-fetch/index").Request'.
              Type 'Request' is not assignable to type 'string | Request'.
                Type 'Request' is not assignable to type 'import("/usr/src/app/node_modules/@types/node-fetch/index").Request'.
                  Property 'context' is missing in type 'Request'.

220       options
          ~~~~~~~
```

### CODE REVIEW

**Comment by cloudify** [comment]
@lussoluca you need to use [this branch of node-fetch](https://github.com/teamdigitale/italia-ts-commons/blob/master/package.json#L35)

**Comment by gunzip** [comment]
the problem is that for some odd reason the fetch definition gets taken from the lib.dom.d.ts typescript definition which has the input argument marked as optional:

`declare function fetch(input?: Request | string, init?: RequestInit): Promise<Response>;`

I don't know why this native delaration overrides the one [here](https://github.com/teamdigitale/italia-ts-commons/blob/master/src/requests.ts#L12), I'm looking at it

**Comment by gunzip** [comment]
@lussoluca as a workaround you can use the following:

```ts
import nodeFetch from "node-fetch";
...

export function APIClient(
  baseUrl: string,
  token: string,
  // tslint:disable-next-line
  fetchApi: typeof fetch = (nodeFetch as any) as typeof fetch
): {
...

```

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>


New dependencies added: <a href='https://www.npmjs.com/package/node-fetch'><code>node-fetch</code></a>.


## Affected stories

  * ⚙️ [#157768307](https://www.pivotaltracker.com/story/show/157768307): Decidere l'approccio da usare per il client REST



<h2><a href="https://github.com/bitinn/node-fetch">node-fetch</a></h2>
<p>Author: David Frank</p>
<p>Description: A light-weight module that brings window.fetch to node.js</p>
<p>Homepage: <a href="https://github.com/bitinn/node-fetch">https://github.com/bitinn/node-fetch</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>about 1 month ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>2</td></tr><tr><td>Releases</td><td>44</td></tr><tr><td>Direct Dependencies</td><td></td></tr><tr><td>Keywords</td><td>fetch, http and promise</td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/clients/api.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+type OcpApimSubscriptionKey = t.TypeOf<typeof OcpApimSubscriptionKey>;
+
+// ProfileLimitedOrExtended is oneOf [LimitedProfile, ExtendedProfile]
+const ProfileLimitedOrExtended = t.union([LimitedProfile, ExtendedProfile]);
+
+export type ProfileLimitedOrExtended = t.TypeOf<
+  typeof ProfileLimitedOrExtended
+>;
+
+export type BasicResponseTypeWith401<R> =
```
Comment (cloudify) [inline src/clients/api.ts]: `BasicResponseType` is enough, the CD API never responds with a 401 (see [the specs](https://github.com/teamdigitale/digital-citizenship-functions/blob/master/api/public_api_v1.yaml))

File: src/clients/api.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  token: string
+): RequestHeaderProducer<P, OcpApimSubscriptionKey> {
+  return () => ({
+    [OcpApimSubscriptionKey]: token
+  });
+}
+
+export type GetProfileT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
```
Comment (cloudify) [inline src/clients/api.ts]: you can actually use `FiscalCode` here, instead of `string`

File: src/clients/api.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<ProfileLimitedOrExtended>
+>;
+
+export type CreateOrUpdateProfileT = IPostApiRequestType<
+  {
+    readonly fiscalCode: string;
```
Comment (cloudify) [inline src/clients/api.ts]: you can actually use `FiscalCode` here, instead of `string`

File: src/clients/api.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    readonly newProfile: ExtendedProfile;
+  },
+  OcpApimSubscriptionKey | "Content-Type",
+  never,
+  BasicResponseTypeWith401<ExtendedProfile>
+>;
+
+export type GetServicesByRecipientT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
```
Comment (cloudify) [inline src/clients/api.ts]: you can actually use `FiscalCode` here, instead of `string`

File: src/clients/api.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  BasicResponseTypeWith401<ServicePublic>
+>;
+
+export function APIClient(
+  baseUrl: string,
+  token: string,
+  // tslint:disable-next-line:no-any
+  fetchApi: typeof fetch = (nodeFetch as any) as typeof fetch
+): {
+  readonly createOrUpdateProfile: (
```
Comment (cloudify) [inline src/clients/api.ts]: you can use [TypeOfApiCall](https://github.com/teamdigitale/italia-ts-commons/blob/master/src/requests.ts#L228):

```ts
readonly createOrUpdateProfile: TypeOfApiCall<CreateOrUpdateProfileT>
```

File: src/clients/api.ts (hunk 1 above)
Comment (gunzip) [reply src/clients/api.ts]: what's the status in case the API gateway cannot authenticate the request ?

Comment (cloudify) [reply src/clients/api.ts]: ah yes, is it supposed to be defined in the specs?

File: src/clients/api.ts (hunk 2 above)
Comment (lmorelli986) [reply src/clients/api.ts]: Fixed from [6f74868] on

File: src/clients/api.ts (hunk 3 above)
Comment (lmorelli986) [reply src/clients/api.ts]: Fixed from [6f74868] on

File: src/clients/api.ts (hunk 4 above)
Comment (lmorelli986) [reply src/clients/api.ts]: Fixed from [6f74868] on

File: src/clients/api.ts (hunk 5 above)
Comment (lmorelli986) [reply src/clients/api.ts]: Ok, thank you. Refactored.

**Pull request merged**